src/consul.rs

```diff
@@ -0,0 +1,452 @@
+
+            consul_index = response
+                .headers()
+                .get("X-Consul-Index")
+                .and_then(|value| value.to_str().ok())
+                .map(|value| value.to_string());
+
+            let kvs = response.json::<Vec<ConsulKVResponse>>().await?;
+
+            for kv in kvs {
+                if kv.key == CONSUL_STORE_KEY && kv.session.is_none() {
+                    println!("=====> lock is free, returning");
+                    return Ok(());
+                }
+            }
+        }
+    }
+
+    /// Retrieves a list of all registered services and parses their tags into DnsTag
+    pub async fn fetch_service_tags(
+        &self,
+        consul_index: &mut Option<String>,
+    ) -> Result<Vec<DnsRecord>, anyhow::Error> {
+        let services_url = self.catalog_api_base_url.join("services")?;
+
+        let mut req = self.http_client.get(services_url);
+
+        if let Some(index) = consul_index {
+            req = req.query(&[("index", &index.to_string())]);
+        }
+
+        // Add a filter to only match "normal" Consul services
+        req = req.query(&[(
+            "filter",
+            r#"ServiceKind == "" and ServiceTags contains "external-dns.enable=true""#,
+        )]);
+
+        let response = req.send().await?.error_for_status()?;
+
+        if let Some(index_header) = response.headers().get("X-Consul-Index") {
+            if let Ok(index_str) = index_header.to_str() {
+                let _ = consul_index.insert(index_str.to_string());
+            } else {
+                eprintln!("Failed to convert HeaderValue to string");
+            }
+        }
+
+        let records = response.json::<HashMap<String, Vec<String>>>().await?;
+
+        let dns_tags = records
+            .into_iter()
+            .flat_map(|(_service_name, tags)| parse_dns_tags(tags))
+            .collect();
+
+        Ok(dns_tags)
+    }
+
+    // Store all DNS records under a single key as a HashMap
+    pub async fn update_consul_dns_records(
+        &self,
+        dns_state: HashMap<String, DnsRecord>,
+    ) -> Result<()> {
+        let url = self.kv_api_base_url.join(CONSUL_STORE_KEY)?;
+
+        self.http_client
+            .put(url)
+            .json(&dns_state)
+            .send()
+            .await?
+            .error_for_status()?;
+        Ok(())
+    }
+
+    /// Fetches all DNS records from Consul.
+    /// This function retrieves the state of all DNS records stored under a specific Consul key.
+    pub async fn fetch_all_dns_records(&self) -> Result<HashMap<String, DnsRecord>, anyhow::Error> {
+        let url = self.kv_api_base_url.join(CONSUL_STORE_KEY)?;
+
+        let resp = self.http_client.get(url).send().await?;
+
+        if !resp.status().is_success() {
+            if resp.status() == StatusCode::NOT_FOUND {
+                return Ok(HashMap::new());
+            }
+            return Err(anyhow::anyhow!(resp.status()));
```

Comment:
Minor thing, but this could be

```suggestion
            anyhow::bail!(resp.status());
```

Reply:
fixed